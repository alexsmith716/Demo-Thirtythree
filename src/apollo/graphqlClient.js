import fetch from 'isomorphic-fetch';

const graphqlClient = ({ endpoint, query, variables={}, method='POST' }) => {

	//	console.log('>>>> graphqlClient > endpoint: ', endpoint);
	//	console.log('>>>> graphqlClient > query: ', query);
	//	console.log('>>>> graphqlClient > variables: ', {...variables});
	//	console.log('>>>> graphqlClient > method: ', method);

	return fetch(endpoint, {
		method: method,
		headers: {
			'Content-Type': 'application/json',
		},
		body: JSON.stringify({
			query: query,
			variables: {...variables}
		}),
	})
		.then(response => response.json())
		.then(data => {
			if (data.error) {
				throw new Error(data.error);
			}
			console.log('>>>> graphqlClient > DATA:', JSON.stringify(data));
			return data;
		})
		.catch(err => {
			console.error('>>>> graphqlClient > ERROR:', err.message);
			return null;
		});
};

export default graphqlClient;
