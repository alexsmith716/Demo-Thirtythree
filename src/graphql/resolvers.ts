
const { KaplanTestPrepBooks } = require('./datasources/googleBooksAPI');
//import { KaplanTestPrepBooks } from './datasources/googleBooksAPI';
const { RickAndMortyAPICharacter } = require('./datasources/rickAndMortyAPI');

export const dataSources = () => ({
	kaplanTestPrepBooks: new KaplanTestPrepBooks(),
	rickAndMortyAPICharacter: new RickAndMortyAPICharacter(),
});

//	https://github.com/dotansimha/graphql-code-generator
export const resolvers = {
	Query: {
		hello: () => 'Hello world!',
		character: async (obj: any, { id }: { id: number }, { dataSources }: { dataSources: any }) => (
			dataSources.rickAndMortyAPICharacter.character({ id })
		),
		search: async (obj: any, { searchString }: { searchString: string }, { dataSources }: { dataSources: any }) => (
			dataSources.kaplanTestPrepBooks.getBooks(searchString)
		),
	},

	Book: {
		id: ({ id }: { id: string }) => id,
		title: ({ volumeInfo: { title } }: { volumeInfo: any, title: string }) => title,
	},

	SearchResult: {
		books: ({ items }: { items: any }) => items,
	}
};
