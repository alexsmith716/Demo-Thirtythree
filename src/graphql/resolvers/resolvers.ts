import  { KaplanTestPrepBooks } from '../datasources/googleBooksAPI';
import  { RickAndMortyAPICharacter } from '../datasources/rickAndMortyAPI';

export const dataSources = () => ({
	kaplanTestPrepBooks: new KaplanTestPrepBooks(),
	rickAndMortyAPICharacter: new RickAndMortyAPICharacter(),
});

export const resolvers = {
	Query: {
		hello: () => 'Hello world!',
		character: async (obj: any, { id }: { id: number }, { dataSources }: { dataSources: any }) => (
			dataSources.rickAndMortyAPICharacter.character({ id })
		),
		search: async (obj: any, { searchString, startIndex, orderBy, maxResults }: { searchString: string, startIndex: number, orderBy: string, maxResults: number }, { dataSources }: { dataSources: any }) => (
			dataSources.kaplanTestPrepBooks.getBooks(searchString, startIndex, orderBy, maxResults)
		),
	},

	Book: {
		id: ({ id }: { id: string }) => id,
		title: ({ volumeInfo: { title }}: { volumeInfo: any, title: string }) => title,
		subtitle: ({ volumeInfo: { subtitle }}: { volumeInfo: any, subtitle: string }) => subtitle,
		authors: ({ volumeInfo: { authors }}: { volumeInfo: any, authors: string }) => authors,
		publisher: ({ volumeInfo: { publisher }}: { volumeInfo: any, publisher: string }) => publisher,
		publishedDate: ({ volumeInfo: { publishedDate }}: { volumeInfo: any, publishedDate: string }) => publishedDate,
		description: ({ volumeInfo: { description }}: { volumeInfo: any, description: string }) => description,
		previewLink: ({ volumeInfo: { previewLink }}: { volumeInfo: any, previewLink: string }) => previewLink,
		imageLinks: ({ volumeInfo: { imageLinks }}: { volumeInfo: any, imageLinks: string }) => imageLinks,
	},

	ImageLinks: {
		smallThumbnail: ({ smallThumbnail }: { smallThumbnail: string }) => smallThumbnail,
		thumbnail: ({ thumbnail }: { thumbnail: string }) => thumbnail,
	},

	SearchResult: {
		books: ({ items }: { items: any }) => items,
	}
};
