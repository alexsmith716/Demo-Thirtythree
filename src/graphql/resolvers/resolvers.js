import { paginateResults } from '../utils/utils';
import { GoogleBooksAPI } from '../datasources/googleBooksAPI';
import { RickAndMortyAPI } from '../datasources/rickAndMortyAPI';
import graphqlClient from '../../apollo/graphqlClient';

// https://github.com/ardatan/graphql-tools
// https://github.com/ardatan/graphql-tools/blob/master/website/docs/generate-schema.md
// https://www.apollographql.com/docs/apollo-server/schema/schema/#the-schema-definition-language
// https://github.com/apollographql/apollo-server/blob/docs/source/api/graphql-tools.md

// Every resolver in a GraphQL.js schema accepts four positional arguments:
//		fieldName(obj, args, context, info) { result }

// These arguments have the following meanings and conventional names:
//		* obj: The object that contains the result returned from the resolver on the parent field, or, in the case of a top-level Query field, the rootValue passed from the server configuration. This argument enables the nested nature of GraphQL queries.
//		* args: An object with the arguments passed into the field in the query. For example, if the field was called with author(name: "Ada"), the args object would be: { "name": "Ada" }.
//		* context: This is an object shared by all resolvers in a particular query, and is used to contain per-request state, including authentication information, dataloader instances, and anything else that should be taken into account when resolving the query. If you're using express-graphql, read about how to set the context in the setup documentation.
//		* info: This argument should only be used in advanced cases, but it contains information about the execution state of the query, including the field name, path to the field from the root, and more. It's only documented in the GraphQL.js source code.

import { GET_RICK_AND_MORTY_CHARACTER, GET_RICK_AND_MORTY_CHARACTERS, GET_RICK_AND_MORTY_CHARACTERS_BY_IDS, } from '../queries/queries';

export const dataSources = () => ({
	googleBooks: new GoogleBooksAPI(),
	rickAndMorty: new RickAndMortyAPI(),
});

// https://github.com/apollographql/apollo-client/blob/docs/source/data/error-handling.mdx

// Resolver functions take the following positional arguments, in order:
// parent, args, context, info
export const resolvers = {
	Query: {
		hello: () => 'Hello world!',

		googleBooks: async (obj, { after, searchString, orderBy, pageSize = 2, }, { dataSources }) => {
			try {
				const allGoogleBooks = await dataSources.googleBooks.getBooks(searchString, orderBy);
				const books = paginateResults({ after, pageSize, results: allGoogleBooks });

				return {
					books,
					cursor: books.length ? books[books.length - 1].cursor : null,
					hasMore: books.length
						? books[books.length - 1].cursor !==
							allGoogleBooks[allGoogleBooks.length - 1].cursor
						: false,
				};
			} catch (error) {
				console.error('>>>>>>>>>>>>> RESOLVERS > Query > googleBooks > ERROR: ', error);
				return false;
			}
		},

		googleBook: (obj, { id }, { dataSources }) => {
			const book = dataSources.googleBooks.getBook({ id });
			return book;
		},

		character: async (obj, { id }, { dataSources }) => {
			try {
				const character = await dataSources.rickAndMorty.getCharacter({ id });
				return character;
			} catch (error) {
				console.error('>>>>>>>>>>>>> RESOLVERS > Query > character > ERROR: ', error);
				return false;
			}
		},

		characters: async (obj, { page, filter }, { dataSources }) => {
			try {
        console.log('#################### RESOLVERS/characters ####################')
				const characters = await dataSources.rickAndMorty.getCharacters({ page, filter });
				return characters;
			} catch (error) {
				console.error('>>>>>>>>>>>>> RESOLVERS > Query > characters > ERROR: ', error);
				return false;
			}
		},

		charactersByIds: async (obj, { ids }, { dataSources }) => {
			try {
				const charactersByIds = await dataSources.rickAndMorty.getCharactersByIds({ ids });
				return charactersByIds;
			} catch (error) {
				console.error('>>>>>>>>>>>>> RESOLVERS > Query > charactersByIds > ERROR: ', error);
				return false;
			}
		},
	},

	// https://github.com/apollographql/apollo-client/blob/main/docs/source/data/mutations.mdx#updating-a-single-existing-entity
	// if mutation updates a single entity, AC automatically updates that value in cache when the mutation returns
	Mutation: {
		googleBookModifyFavorite: async (obj, { id, favorite }, { dataSources }) => {
			try {
				const book = await dataSources.googleBooks.getBook({ id });
				book.favorite = favorite;
				return {
					success: true,
					message: 'added to favorites',
					books: [book],
				};
			} catch (error) {
				console.error('>>>>>>>>>>>>> RESOLVERS > Mutation > googleBookModifyFavorite > ERROR: ', error);
				return false;
			}
		},
	},
};
