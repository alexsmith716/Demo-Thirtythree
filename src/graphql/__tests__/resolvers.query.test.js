//	import { graphql } from 'graphql';
//	import { gql } from '@apollo/client';
//	import { createTestClient } from 'apollo-server-testing';

//	import { constructTestServer, executableSchema } from '../executableSchema';
//	import { GET_RICK_AND_MORTY_CHARACTER, } from '../queries/queries.js';
import { resolvers } from '../resolvers/resolvers';
import graphqlClient from '../../apollo/graphqlClient';

// =====================================================

// https://api.spacexdata.com/v2/launches?flight_number=99

//	graphql(
//		schema: GraphQLSchema,
//		requestString: string,
//		rootValue?: ?any,
//		contextValue?: ?any,
//		variableValues?: ?{[key: string]: any},
//		operationName?: ?string
//	): Promise<GraphQLResult>

// The graphql function lexes, parses, validates and executes a GraphQL request.
// It requires a schema and a requestString.
// Optional arguments include a rootValue, which will get passed as the root value to the executor, 
//		a contextValue, which will get passed to all resolve functions, variableValues, 
//		which will get passed to the executor to provide values for any variables in requestString, 
//		and operationName, which allows the caller to specify which operation in requestString will be run, 
//		in cases where requestString contains multiple top-level operations.

// =====================================================

describe('[Query.character]', () => {
	const mockContext = {
		dataSources: {
			rickAndMorty: { getCharacter: jest.fn() },
		},
	};

	it('calls lookup from rickAndMorty api', async () => {
		const getCharacter = mockContext.dataSources.rickAndMorty.getCharacter;
		getCharacter.mockReturnValueOnce({id: 11});

		// check the resolver response
		const res = await resolvers.Query.character(null, { id: 11 }, mockContext);
		expect(res).toEqual({id: 11});

		expect(getCharacter).toBeCalledWith({ id: 11 });
		expect(res).toMatchSnapshot();
	});
});

// =====================================================

describe('[Query.googleBook]', () => {
	const mockContext = {
		dataSources: {
			googleBooks: { getBook: jest.fn() },
		},
	};

	it('calls lookup from googleBook api', async () => {
		const getBook = mockContext.dataSources.googleBooks.getBook;
		getBook.mockReturnValueOnce({ id: 'uW_zzQEACAAJ' });

		// check the resolver response
		const res = await resolvers.Query.googleBook(null, { id: 'uW_zzQEACAAJ' }, mockContext);
		expect(res).toEqual({ id: 'uW_zzQEACAAJ' });

		// make sure the dataSources were called properly
		expect(getBook).toBeCalledWith({ id: 'uW_zzQEACAAJ' });
		expect(res).toMatchSnapshot();
	});
});

// =====================================================

describe('[Query.googleBooks]', () => {
	const mockContext = {
		dataSources: {
			googleBooks: { getBooks: jest.fn() },
		},
	};

	const { getBooks } = mockContext.dataSources.googleBooks;

	it('calls lookup from googleBooks api', async () => {
		getBooks.mockReturnValueOnce([{ id: 'uW_zzQEACAAJ', cursor: 'foo' }]);

		// check the resolver response
		const res = await resolvers.Query.googleBooks(null, {}, mockContext);
		expect(res).toEqual({
			books: [{ id: 'uW_zzQEACAAJ', cursor: 'foo' }],
			cursor: 'foo',
			hasMore: false,
		});
		expect(res).toMatchSnapshot();
	});
});
