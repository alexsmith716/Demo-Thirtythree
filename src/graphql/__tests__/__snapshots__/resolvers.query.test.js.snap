// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Query.character] calls lookup from rickAndMorty api 1`] = `
Object {
  "id": 11,
}
`;

exports[`[Query.googleBook] calls lookup from googleBook api 1`] = `
Object {
  "id": "uW_zzQEACAAJ",
}
`;

exports[`[Query.googleBooks] calls lookup from googleBooks api 1`] = `
Object {
  "books": Array [
    Object {
      "cursor": "foo",
      "id": "uW_zzQEACAAJ",
    },
  ],
  "cursor": "foo",
  "hasMore": false,
}
`;
